frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::{SortedListProvider, VoteWeight};
+	use frame_support::pallet_prelude::*;
+
+	use sp_staking::StakingInterface;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// This has to come from Staking::Currency
+		type Currency: Currency<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// A sorted list of nominators and validators, by their stake and self-stake respectively.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn active_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Staking::stake(&who).map(|s| s.active).unwrap_or_default()
+	}
+
+	pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+		let total_issuance = T::Currency::total_issuance();
+		<T::Staking as StakingInterface>::CurrencyToVote::to_vote(balance, total_issuance)
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn on_stake_update(who: &T::AccountId, _: Option<Stake<T::AccountId, BalanceOf<T>>>) {
+		if let Ok(current_stake) = T::Staking::stake(who) {
+			let current_active = current_stake.active;
+
+			// if this is a nominator
+			if let Some(_) = T::Staking::nominations(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+
+			if T::Staking::is_validator(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active));
+			}
+		}
+	}
+
+	fn on_nominator_update(who: &T::AccountId, _prev_nominations: Vec<T::AccountId>) {
+		// NOTE: We ignore the result here, because this method can be called when the nominator
+		// is already in the list, just changing their nominations.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(Self::active_stake_of(who)));
+	}
+
+	fn on_validator_update(who: &T::AccountId) {
+		let self_stake = Self::active_stake_of(who);
+		// maybe update sorted list.
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(self_stake));
+	}
+
+	fn on_validator_remove(who: &T::AccountId) {
+		let _ = T::VoterList::on_remove(who);
+	}
+
+	fn on_nominator_remove(who: &T::AccountId, _nominations: Vec<T::AccountId>) {
+		let _ = T::VoterList::on_remove(who);
+	}
+
+	fn on_unstake(_who: &T::AccountId) {}
+}
+
+/// A wrapper for a given `SortedListProvider` that introduces defensive checks  for insert, update
+/// and remove operations, suggesting that it's read-only, except for unsafe operations.
+pub struct TrackedList<T, S, P>(sp_std::marker::PhantomData<(T, S, P)>);
```

Comment:
Don't we actually want to call this `struct ReadOnly`?

Reply:
Also, I think

```
struct ReadOnly<AccountId, Inner>(..)
```

will suffice, no? 

1. You don't need `T`, only account id.
2. You can always use `Inner::Score`, do we need another generic `S` and bound it to `Inner::Score`, as you do now? I think that's not needed.

Reply:
Since this struct now just does the defensive checks it felt weird to call it ReadOnly, also the name "TrackedList" seems to be lining up nicely with having those defensive checks on certain operations. 
Yes, I think we can restrict the types further as you suggested.

Reply:
Naming up to you, extra types could go away.